http/custom_roundtripper.go

```diff
@@ -0,0 +1,55 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log"
+	"net/http"
+	"time"
+
+	awsAuth "github.com/ONSdigital/dp-elasticsearch/v2/awsauth"
+)
+
+type awsSignerRoundTripper struct {
+	signer       *awsAuth.Signer
+	roundTripper http.RoundTripper
+}
+
+func NewAWSSignerRoundTripper(awsFilename, awsProfile, awsRegion, awsService string, customTransport http.RoundTripper) http.RoundTripper {
+	var roundTripper http.RoundTripper
+	if awsRegion == "" || awsService == "" {
+		log.Fatal(context.Background(), "aws region and service should be valid options")
+	}
+	awsSigner, err := awsAuth.NewAwsSigner(awsFilename, awsProfile, awsRegion, awsService)
+	if err != nil {
+		log.Fatal(context.Background(), "failed to create aws v4 signer", err)
+	}
+
+	if customTransport == nil {
+		roundTripper = http.DefaultTransport
+	} else {
+		roundTripper = customTransport
+	}
+
+	return &awsSignerRoundTripper{
+		signer:       awsSigner,
+		roundTripper: roundTripper,
+	}
+}
+
+func (srt *awsSignerRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	var body []byte
+	var err error
+	if req.Body != nil {
+		body, err = io.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
```

Comment:
A note on returning errors - it's almost always better to wrap the error in some sort of human readable context explaining what went wrong, i.e. here:
`return nil, fmt.Errorf("failed to read request body: %w", err)`
will be a lot more informative if an error is ever thrown. A random `EOF` can easily send you on a wild goose chase, but `failed to read request body: EOF` is already a lot better.
The same for below (line 50) and really any time returning an error unless you're certain it would be redundant.

Reply:
This is a library that will have lots of consumers. So each consumer can use this library and wrap the errors they have received with human readable errors using the wrap . This library won't be exposed directly to the outside world. So it should be fine i guess to return the actual erros in here and let the consumers wrap with more meaningful messages.,

Reply:
I'm not sure I agree with that logic.. If one of these errors is returned with an ambiguous or similar message it will be hard to debug. Callers should also be wrapping their errors with further context, but as a user of the library I'd like to know if the request body has failed to be read or the signing has failed.

Reply:
Again this is an underlying custom implementation of aws signer roundtripper. As a consumer , I think i only need to know if the signing is succeeded or not, I don't think it should be wrapped to further explain it . If this is an intermittent error or retrying helps etc then there needs wrapping but other than that my personal take is that here there is no wrapping needed . But if you are really strong on this then happy to change this one,

Reply:
I'm thinking more in terms of debugging - this is a package that will be used mostly by us, so if we're working with it and something goes wrong it's always going to be better to know exactly where and why. I don't I've ever regretted having too much information in an error, unless there's repetition or sensitive information exposed. Yes I think it should be changed, thanks.

Reply:
let me know your thoughts on this and i can change it .